### PR TITLE
Optimize graph construction pipeline

### PR DESCRIPTION
**Pull request by zheng-da**
*Description of changes:*
The graph construction pipeline cannot handle large graphs. This PR is to optimize the resource management to enable it to process a graph with 1B nodes and 10B edges.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/gconstruct/construct_graph.py [hunk 1]
```diff
@@ -87,17 +100,30 @@ def write_data_parquet(data, data_file):
     table = pa.Table.from_arrays(list(arr_dict.values()), names=list(arr_dict.keys()))
     pq.write_table(table, data_file)
 
-def parse_file_format(fmt):
+def parse_file_format(conf, is_node):
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:103]: Can we have:
```
def parse_node_file_format(conf):
    _parse_file_format(conf, is_node=True)

def parse_edge_file_format(conf):
    _parse_file_format(conf, is_node=False)

def _parse_file_format(conf, is_node):
   ...
```
The helps people better under stand the main function.

File: python/graphstorm/gconstruct/construct_graph.py [hunk 2]
```diff
@@ -87,17 +100,30 @@ def write_data_parquet(data, data_file):
...
 
     Parameters
     ----------
-    fmt : dict
-        Describe the file format.
+    conf : dict
+        Describe the config for the node type or edge type.
+    is_node : bool
+        Whether this is a node config or edge config
+
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:112]: Can we have an example of the config here?
```
Example of conf:
{
  "format": {
    "name": "parquet",
  }
  "node_id_col": ID,
  "features": [feat0, feat1, ...],
  "labels": [label0],
}
```

File: python/graphstorm/gconstruct/construct_graph.py [hunk 3]
```diff
@@ -357,30 +354,29 @@ def parse_node_data(file_idx, in_file, feat_ops, node_id_col, label_conf,
...
-                    node_id_map, label_conf, read_file, return_dict):
+def parse_edge_data(in_file, feat_ops, src_id_col, dst_id_col, edge_type,
+                    node_id_map, label_conf, read_file):
     """ Parse edge data.
 
     The function parses an edge file that contains the source and destination node
     IDs, edge features and potentially edge labels. The edge file is parsed
-    according to users' configuration and performs some feature transformation
-    and save the result in `return_dict`.
+    according to users' configuration and performs some feature transformation.
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:376]: Can we tell what kinds of feature transformation GFS supports?

File: python/graphstorm/gconstruct/construct_graph.py [hunk 4]
```diff
@@ -438,6 +436,86 @@ def create_id_map(ids):
...
+    user_parser : callable
+        The user-defined function to read and process the data files.
+    """
+    try:
+        while True:
+            # If the queue is empty, it will raise the Empty exception.
+            i, in_file = task_queue.get_nowait()
+            data = user_parser(in_file)
+            res_queue.put((i, data))
+            gc.collect()
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:458]: Why we need a gc collect?

File: python/graphstorm/gconstruct/construct_graph.py [hunk 5]
```diff
@@ -438,6 +436,86 @@ def create_id_map(ids):
...
+    num_processes : int
+        The number of processes that run in parallel.
+    user_parser : callable
+        The user-defined function to read and process the data files.
+    """
+    def __init__(self, name, in_files, num_processes, user_parser):
+        self.name = name
+        self.processes = []
+        manager = multiprocessing.Manager()
+        self.task_queue = manager.Queue()
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:488]: Can we make the size of res_queue tunable? Or 8 is good enough?

File: python/graphstorm/gconstruct/construct_graph.py [hunk 6]
```diff
@@ -496,23 +574,24 @@ def process_node_data(process_confs, remap_id, num_processes):
         node_type = process_conf['node_type']
         assert 'format' in process_conf, \
                 "'format' must be defined for a node type"
-        read_file = parse_file_format(process_conf['format'])
+        read_file = parse_file_format(process_conf, True)
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:577]:
```suggestion
        read_file = parse_file_format(process_conf, is_node=True)
```
or
```suggestion
        read_file = parse_node_file_format(process_conf)
```

File: python/graphstorm/gconstruct/construct_graph.py [hunk 7]
```diff
@@ -496,23 +574,24 @@ def process_node_data(process_confs, remap_id, num_processes):
...
+
+        user_parser = partial(parse_node_data, feat_ops=feat_ops,
+                              node_id_col=node_id_col,
+                              label_conf=label_conf,
+                              read_file=read_file)
+        start = time.time()
+        pool = WorkerPool(node_type, in_files, num_processes, user_parser)
+        return_dict = pool.get_data()
+        pool.close()
+        print("Processing data files for node {} takes {:.3f} seconds".format(
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:593]:
```suggestion
        print(f"Processing data files for node {node_type} takes {time.time() - start:.3f} seconds"
```

File: python/graphstorm/gconstruct/construct_graph.py [hunk 8]
```diff
@@ -537,17 +619,23 @@ def process_node_data(process_confs, remap_id, num_processes):
         else:
             type_node_id_map = create_id_map(type_node_id_map)
             num_nodes = len(type_node_id_map)
+        sys_tracker.check(f'Create node ID map of {node_type}')
 
         for feat_name in type_node_data:
             type_node_data[feat_name] = np.concatenate(type_node_data[feat_name])
             assert len(type_node_data[feat_name]) == num_nodes
+            print("node type {} has feature {} of {}".format(
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:627]:
```suggestion
            print(f"node type {node_type} has feature {feat_name} " \
                     f"of {type_node_data[feat_name].shape}"
```

File: python/graphstorm/gconstruct/construct_graph.py [hunk 9]
```diff
@@ -609,25 +697,27 @@ def process_edge_data(process_confs, node_id_map, num_processes):
         edge_type = process_conf['relation']
         assert 'format' in process_conf, \
                 "'format' is not defined for an edge type."
-        read_file = parse_file_format(process_conf['format'])
+        read_file = parse_file_format(process_conf, False)
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:700]:
```suggestion
        read_file = parse_file_format(process_conf, is_node=False)
```
or
```suggestion
        read_file = parse_edge_file_format(process_conf)
```

File: python/graphstorm/gconstruct/construct_graph.py [hunk 10]
```diff
@@ -609,25 +697,27 @@ def process_edge_data(process_confs, node_id_map, num_processes):
...
+                              dst_id_col=dst_id_col,
+                              edge_type=edge_type,
+                              node_id_map=node_id_map,
+                              label_conf=label_conf,
+                              read_file=read_file)
+        start = time.time()
+        pool = WorkerPool(edge_type, in_files, num_processes, user_parser)
+        return_dict = pool.get_data()
+        pool.close()
+        print("Processing data files for edges of {} takes {:.3f} seconds".format(
```
Comment (classicsong) [inline python/graphstorm/gconstruct/construct_graph.py:719]: Use print(f"")

**Review by classicsong — APPROVED** [review]
Overall it looks good.

**Comment by classicsong** [comment]
Can we add some more tests code about 1) loading multiple node files and 2) loading multiple edge files?

File: python/graphstorm/gconstruct/construct_graph.py (hunk 5 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:488]: we probably don't need to tune that much. let's set a default number for users first. otherwise, there will be too many arguments for users to understand.

File: python/graphstorm/gconstruct/construct_graph.py (hunk 4 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:458]: i try to trigger garbage collection as much as possible.

File: python/graphstorm/gconstruct/construct_graph.py (hunk 3 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:376]: This may not be the best place to document the transformation the pipeline supports. We need to have proper documentation for the pipeline.

File: python/graphstorm/gconstruct/construct_graph.py (hunk 2 above)
Comment (zheng-da) [reply python/graphstorm/gconstruct/construct_graph.py:112]: The config has been documented in `process_node_data` and `process_edge_data`.

**Comment by zheng-da** [comment]
There is an integration test that processes multiple node types with multiple node data files and multiple edge types with multiple edge data files. It tests all different kinds of combinations.
